interp/constrexpr.ml

```diff
@@ -48,18 +48,41 @@ type abstraction_kind = AbsLambda | AbsPi
 
 type proj_flag = int option (** [Some n] = proj of the n-th visible argument *)
 
-(** Representation of integer literals that appear in Coq scripts.
-    We now use raw strings of digits in base 10 (big-endian), and a separate
-    sign flag. Note that this representation is not unique, due to possible
-    multiple leading zeros, and -0 = +0 *)
+(** Representation of decimal literals that appear in Coq scripts.
+   We now use raw strings of digits in base 10 (big-endian),
+   a separate sign flag (true meaning + and false meaning -), an
+   optional fractional part (same format as the mandatory integral
+   part) and an optional exponent. E.g., Numeral (false, "1", Some
+   ("023"), Some (Expep, "02")) means -1.023e+02 (i.e. -102.3).
+
+   Note that this representation is not unique, due to possible
+   multiple leading or trailing zeros, and -0 = +0, for instances.
+   The goal is to ensure that numerals are printed exactly as they are
+   parsed (otherwise, they would become unusable as notations (for
+   instance, "00" could not be used as notation if it is interpreted
+   as "0")). *)
 
 type sign = bool
-type raw_natural_number = string
+type raw_natural_number = string  (** \[0-9\]+ *)
+type exp_tag = Expe | Expep | Expem | ExpE | ExpEp | ExpEm
+
+let string_of_exp_tag = function
+  | Expe -> "e" | Expep -> "e+" | Expem -> "e-"
+  | ExpE -> "E" | ExpEp -> "E+" | ExpEm -> "E-"
 
 type prim_token =
-  | Numeral of raw_natural_number * sign
+  | Numeral of sign * raw_natural_number  (** integral part *)
+               * (raw_natural_number) option  (** fractional part *)
+               * (exp_tag * raw_natural_number) option  (** decimal exponent *)
```

Comment:
More generally, let me make explicit that refining half-way the structure of `Numeral` as you do is somehow strange, compared to do the decomposition in the `coqdecimal_of_rawnum` function. But why not.